File: Hands of Straights - GFG/hands-of-straights.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.math.*;
import java.io.*;

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            int n = sc.nextInt();
            int k = sc.nextInt();
            int a[] = new int[n];
            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
            }
            Solution obj = new Solution();
            boolean ans = obj.isStraightHand(n, k, a);
            System.out.println(ans ? "True" : "False");
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    static boolean isStraightHand(int n, int gs, int arr[]) {
        // code here
        if (n % gs != 0) return false;
        Arrays.sort(arr);
        Map<Integer, Integer> mp = new HashMap<>();
        for (int it : arr) mp.put(it, mp.getOrDefault(it, 0) + 1);
        for (int it : arr) {
            int cnt = gs;
            int val = it;
            while (mp.containsKey(val) && cnt > 0) {
                mp.put(val, mp.get(val) - 1);
                cnt--;
                if (mp.get(val) == 0) mp.remove(val);
                val++;
            }
            if (cnt != 0 && cnt != gs) return false;
        }
        return true;
    }
}

File: Hands of Straights - GFG/hands-of-straights.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    bool isStraightHand(int n, int gs, vector<int> &arr) {
        // code here
        if(n % gs != 0) return false;
        sort(arr.begin(), arr.end());
        unordered_map<int, int> mp;
        for(int it : arr) mp[it]++;
        for(int it : arr) {
            int cnt = gs, val = it;
            while(mp.find(val) != mp.end() && cnt > 0) {
                mp[val] -= 1;
                cnt -= 1;
                if(mp[val] == 0) mp.erase(val);
                val += 1;
            }
            if(cnt != 0 && cnt != gs) return false;
        }
        return true;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, groupSize;
        cin >> N >> groupSize;

        vector<int> hand(N);
        for (int i = 0; i < N; i++) cin >> hand[i];

        Solution obj;
        int ans = obj.isStraightHand(N, groupSize, hand);
        if (ans)
            cout << "True" << endl;
        else
            cout << "False" << endl;
    }
    return 0;
}
// } Driver Code Ends